Proposal: Add real-time priorities for processes

duffsterlp opened the pull request:
This is a proposal to add a hook into docker to allow processes to run with realtime priorities within containers. This is my first post, and I don't know too much about the docker source code, but doing a cursory glance through the source code, I was thinking that this could be implemented by altering the cpu.rt_runtime_us parameter within the cgroup hierarchy. This idea is based on the following blog post:
http://www.breakage.org/2014/08/22/using-sched_fifo-in-docker-containers-on-rhel/

Based on the blog post, it seems feasible to programmatically alter cpu.rt_runtime_us within docker-[id].scope (/sys/fs/cgroup/cpu/system.slice/docker-[long_id].scope/cpu.rt_runtime_us). There is already code within libcontainer to modify other parameters such as CPU shares. My thought was that if the proposed parameter was set to true, it would set rt_runtime_us to 950000 (the default within /proc/sched/kernel/sched_rt_runtime_us). If the proposed parameter was set to false, it would set rt_runtime_us to 0. 

The piece that is less obvious to me is whether Docker should be responsible for altering the parameter within the parent cgroup: (/sys/fs/cgroup/cpu/system.slice/cpu.rt_runtime_us), since this may have an effect on other processes within that slice. 

One thing that's not completely clean about my approach is that the sys_nice capability needs to be added in conjunction with my proposed parameter being set to true. This could be added to the documentation of that parameter, or the setting of that parameter to true could do something like automatically add the sys_nice capability. This doesn't seem too bad since other Docker command line arguments are dependent on each other -> for example, you cannot invoke the -d and the --rm option at the same time.

Note that I've seen PR #9437 and this did not solve the issue; it seems necessary to tell the cgroup that it has RT runtime on the CPU, in addition to setting the ulimits properly.

It is worth mentioning that this is simply one idea for this implementation -> if folks know how to accomplish this using a different method, please feel free to bring it up.

In testing this feature, one can use the chrt utility as described in the blog post. Or, one can invoke the condition that failed for me in the first place that prompted this feature: the function sched_setscheduler() in C. The following C program demonstrates the test condition:

```
#include <stdio.h>
#include <sched.h>

int main()
{
    struct sched_param param;
    param.sched_priority = 50;
    const int myself  =  0; // 0 is the PID of ourself
    if (0 != sched_setscheduler(myself, SCHED_FIFO, &param))
    {
        printf("Failure\n");
        return -1;
    }

    printf("Success\n");
    return 0;
}
```

Information about my environment:

[root@localhost cpu]# cat /etc/*-release
CentOS Linux release 7.0.1406 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CentOS Linux release 7.0.1406 (Core)
CentOS Linux release 7.0.1406 (Core)

[root@localhost cpu]# docker version
Client version: 1.3.2
Client API version: 1.15
Go version (client): go1.3.3
Git commit (client): 39fa2fa/1.3.2
OS/Arch (client): linux/amd64
Server version: 1.3.2
Server API version: 1.15
Go version (server): go1.3.3
Git commit (server): 39fa2fa/1.3.2

Signed-off-by: Glenn Duffy glennduffy@gmail.com

jfrazelle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

duffsterlp commented:
I added the signed off line at the bottom of my proposal. Is that sufficient?

unclejack commented:
@duffsterlp No, that's not enough. You have to add it to the commit itself.

However, signing the commit doesn't mean that it's automatically going to be merged, it still needs to be reviewed and approved for that to happen.

duffsterlp committed:
Proposal: Add real-time priorities for processes

This is a proposal to add a hook into docker to allow processes to run with realtime priorities within containers. This is my first post, and I don't know too much about the docker source code, but doing a cursory glance through the source code, I was thinking that this could be implemented by altering the cpu.rt_runtime_us parameter within the cgroup hierarchy. This idea is based on the following blog post:
http://www.breakage.org/2014/08/22/using-sched_fifo-in-docker-containers-on-rhel/

Based on the blog post, it seems feasible to programmatically alter cpu.rt_runtime_us within docker-[id].scope (/sys/fs/cgroup/cpu/system.slice/docker-[long_id].scope/cpu.rt_runtime_us). There is already code within libcontainer to modify other parameters such as CPU shares. My thought was that if the proposed parameter was set to true, it would set rt_runtime_us to 950000 (the default within /proc/sched/kernel/sched_rt_runtime_us). If the proposed parameter was set to false, it would set rt_runtime_us to 0.

The piece that is less obvious to me is whether Docker should be responsible for altering the parameter within the parent cgroup: (/sys/fs/cgroup/cpu/system.slice/cpu.rt_runtime_us), since this may have an effect on other processes within that slice.

One thing that's not completely clean about my approach is that the sys_nice capability needs to be added in conjunction with my proposed parameter being set to true. This could be added to the documentation of that parameter, or the setting of that parameter to true could do something like automatically add the sys_nice capability. This doesn't seem too bad since other Docker command line arguments are dependent on each other -> for example, you cannot invoke the -d and the --rm option at the same time.

Note that I've seen PR #9437 and this did not solve the issue; it seems necessary to tell the cgroup that it has RT runtime on the CPU, in addition to setting the ulimits properly.

It is worth mentioning that this is simply one idea for this implementation -> if folks know how to accomplish this using a different method, please feel free to bring it up.

In testing this feature, one can use the chrt utility as described in the blog post. Or, one can invoke the condition that failed for me in the first place that prompted this feature: the function sched_setscheduler() in C. The following C program demonstrates the test condition:

int main()
{
    struct sched_param param;
    param.sched_priority = 50;
    const int myself  =  0; // 0 is the PID of ourself
    if (0 != sched_setscheduler(myself, SCHED_FIFO, &param))
    {
        printf("Failure\n");
        return -1;
    }

    printf("Success\n");
    return 0;
}

Information about my environment:

[root@localhost cpu]# cat /etc/*-release
CentOS Linux release 7.0.1406 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CentOS Linux release 7.0.1406 (Core)
CentOS Linux release 7.0.1406 (Core)

[root@localhost cpu]# docker version
Client version: 1.3.2
Client API version: 1.15
Go version (client): go1.3.3
Git commit (client): 39fa2fa/1.3.2
OS/Arch (client): linux/amd64
Server version: 1.3.2
Server API version: 1.15
Go version (server): go1.3.3
Git commit (server): 39fa2fa/1.3.2

Signed-off-by: Glenn Duffy <glennduffy@gmail.com>

duffsterlp commented:
@unclejack Did that do it? Also, is there a way to accomplish signing via the Web GUI alone without using the CLI. Apologies, as I am clearly new to git.

duffsterlp commented:
Also, my C code wasn't parsed correctly by git -> the two includes at the top of the code are <stdio.h> and <sched.h>

thaJeztah commented:
> Also, my C code wasn't parsed correctly by git -> the two includes at the top of the code are stdio.h and sched.h

@duffsterlp you should be able to fix that, either by indenting the entire code sample with 4 spaces or adding a line with three back-ticks (```) before and after the code block.

duffsterlp commented:
@thaJeztah Thanks! Comment is updated.

icecrime commented:
Ping @vishh: what do you think about adding this feature?

vishh commented:
@icecrime: I personally find this feature to be outside of docker's scope since it demands node management and not just that of docker containers. Perhaps `--cgroup-parent` feature discussed in #8551 (PR #11428) can be used to get the desired behavior. A systemd cgroup with appropriate rt scheduling options set can be used a parent cgroup to the docker container.

duffsterlp commented:
@icecrime @vishh Thanks for bringing this up again. 
It is worth noting that after doing some tests, I think it is worth being able to configure the numerical runtime value rather than a boolean. See this stackoverflow issue for reasoning. http://stackoverflow.com/questions/28493333/error-writing-to-cgroup-parameter-cpu-rt-runtime-us/28507639#28507639
Essentially, the default value of 950000 only lets you run a single container because starting another one puts you over the maximum (1000000) for the parent cgroup.

icecrime commented:
Ping @crosbymichael: WDYT?

LK4D4 commented:
@duffsterlp We merged `--parent-cgroup` stuff. Is it will be enough for your usecase? Because I feel little awkward about introducing such flag to run as @vishh too.

icecrime commented:
There are many voices against, so I'm gonna close this for now but feel free to continue the conversation.

duffsterlp commented:
Where are the new flags for the parent cgroup? Is it daemon configuration or in the run command? I don't see a flag called --parent-cgroup anywhere. Also, wherever this new flag is, how can I leverage it for my use case?

vishh commented:
The flag is `--cgroup-parent` and it will be part of docker 1.6. It is a runtime configutation. In your case, you can manage the cgroup hierarchies outside of docker and instruct docker to run the container under a specified cgroup.
